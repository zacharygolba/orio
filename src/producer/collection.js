// @flow

import type { Producer } from './types'

export class MapProducer<K, V> implements Producer<*> {
  size: number
  source: Iterator<[K, V]>
  /*:: @@iterator: () => Iterator<[K, V]> */

  constructor(source: Map<K, V>) {
    this.size = source.size
    this.source = source.entries()
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<[K, V]> {
    return this
  }

  next(): IteratorResult<[K, V], void> {
    return this.source.next()
  }

  sizeHint(): number {
    return this.size
  }
}

export class SetProducer<T> implements Producer<T> {
  size: number
  source: Iterator<T>
  /*:: @@iterator: () => Iterator<T> */

  constructor(source: Set<T>) {
    this.size = source.size
    this.source = source.values()
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<T> {
    return this
  }

  next(): IteratorResult<T, void> {
    return this.source.next()
  }

  sizeHint(): number {
    return this.size
  }
}
