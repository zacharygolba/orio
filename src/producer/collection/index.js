// @flow

export { default as IndexedProducer } from './indexed'
export { default as MapProducer } from './map'
export { default as SetProducer } from './set'
