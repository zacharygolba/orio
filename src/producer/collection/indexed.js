// @flow

import * as result from '../../result'
import ProducerBase from '../base'

export interface IndexedCollection<T> extends Iterable<T> {
  [key: number]: T,
  length: number,
}

export default class IndexedProducer<T> extends ProducerBase<T> {
  source: IndexedCollection<T>
  size: number
  state: number

  constructor(source: IndexedCollection<T>) {
    super()
    this.source = source
    this.size = source.length
    this.state = 0
  }

  next(): IteratorResult<T, void> {
    if (this.state >= this.size) {
      return result.done()
    }

    const value = this.source[this.state]

    this.state += 1
    return result.next(value)
  }

  sizeHint(): number {
    return this.size
  }
}
