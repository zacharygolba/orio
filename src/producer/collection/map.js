// @flow

import type { Producer } from '../types'

export default class MapProducer<K, V> implements Producer<*> {
  size: number
  source: Iterator<[K, V]>
  /*:: @@iterator: () => Iterator<[K, V]> */

  constructor(source: Map<K, V>) {
    this.size = source.size
    this.source = source.entries()
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<[K, V]> {
    return this
  }

  next(): IteratorResult<[K, V], void> {
    return this.source.next()
  }

  sizeHint(): number {
    return this.size
  }
}
