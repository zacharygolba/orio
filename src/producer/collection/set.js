// @flow

import type { Producer } from '../types'

export default class SetProducer<T> implements Producer<T> {
  size: number
  source: Iterator<T>
  /*:: @@iterator: () => Iterator<T> */

  constructor(source: Set<T>) {
    this.size = source.size
    this.source = source.values()
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<T> {
    return this
  }

  next(): IteratorResult<T, void> {
    return this.source.next()
  }

  sizeHint(): number {
    return this.size
  }
}
