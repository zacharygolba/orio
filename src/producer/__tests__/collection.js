// @flow

import { MapProducer, SetProducer } from '../collection'

describe('MapProducer', () => {
  let source
  let producer

  beforeEach(() => {
    source = new Map([[1, 2], [3, 4]])
    producer = new MapProducer(source)
  })

  test('#constructor()', () => {
    expect(producer).toMatchSnapshot()
  })

  test('#@@iterator()', () => {
    for (const item of producer) {
      expect(item).toMatchSnapshot()
    }
  })

  test('#sizeHint()', () => {
    expect(producer.sizeHint()).toBe(source.size)
  })
})

describe('SetProducer', () => {
  let source
  let producer

  beforeEach(() => {
    source = new Set([1, 2, 3])
    producer = new SetProducer(source)
  })

  test('#constructor()', () => {
    expect(producer).toMatchSnapshot()
  })

  test('#@@iterator()', () => {
    for (const item of producer) {
      expect(item).toMatchSnapshot()
    }
  })

  test('#sizeHint()', () => {
    expect(producer.sizeHint()).toBe(source.size)
  })
})
