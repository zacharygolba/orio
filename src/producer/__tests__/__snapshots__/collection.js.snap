// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexedProducer source array #@@iterator() 1`] = `1`;

exports[`IndexedProducer source array #@@iterator() 2`] = `2`;

exports[`IndexedProducer source array #@@iterator() 3`] = `3`;

exports[`IndexedProducer source array #constructor() 1`] = `
IndexedProducer {
  "size": 3,
  "source": Array [
    1,
    2,
    3,
  ],
  "state": 0,
}
`;

exports[`IndexedProducer source string #@@iterator() 1`] = `"t"`;

exports[`IndexedProducer source string #@@iterator() 2`] = `"e"`;

exports[`IndexedProducer source string #@@iterator() 3`] = `"s"`;

exports[`IndexedProducer source string #@@iterator() 4`] = `"t"`;

exports[`IndexedProducer source string #constructor() 1`] = `
IndexedProducer {
  "size": 4,
  "source": "test",
  "state": 0,
}
`;

exports[`MapProducer #@@iterator() 1`] = `
Array [
  1,
  2,
]
`;

exports[`MapProducer #@@iterator() 2`] = `
Array [
  3,
  4,
]
`;

exports[`MapProducer #constructor() 1`] = `
MapProducer {
  "size": 2,
  "source": Object {},
}
`;

exports[`SetProducer #@@iterator() 1`] = `1`;

exports[`SetProducer #@@iterator() 2`] = `2`;

exports[`SetProducer #@@iterator() 3`] = `3`;

exports[`SetProducer #constructor() 1`] = `
SetProducer {
  "size": 3,
  "source": Object {},
}
`;
